File: 5th sem/hw5.py
import numpy as np
import matplotlib.pyplot as plt

# Define constants
n = 51
pi = 3.14159265359

# Initialize parameters
dx = 1.0 / (n - 1)
dt = 0.5 * dx * dx
eps = 0.0000000001
oldP = np.zeros(n)
newP = np.zeros(n)
f = np.zeros(n)
d = np.zeros(n)
alfa = np.zeros(n)
betta = np.zeros(n)

# Set initial conditions
for i in range(n):
    oldP[i] = np.sin(pi * i * dx)

# Set right-hand side
for i in range(n):
    f[i] = 0.0

# Coefficients for the Thomas algorithm
a = -1.0 / (dx * dx)
b = 1.0 / dt + 2.0 / (dx * dx)
c = -1.0 / (dx * dx)

# Main iteration loop
iter = 0
max_diff = float('inf')

while iter < 10000 and max_diff > eps:
    # Fill the d vector
    for i in range(n):
        d[i] = oldP[i] / dt + f[i]

    # Forward substitution (Thomas algorithm)
    alfa[1] = 0.0
    betta[1] = 0.0
    for i in range(1, n - 1):
        alfa[i + 1] = -a / (b + c * alfa[i])
        betta[i + 1] = (d[i] - c * betta[i]) / (b + c * alfa[i])

    # Backward substitution (Thomas algorithm)
    newP[n - 1] = 0.0  # Boundary condition
    for i in range(n - 2, -1, -1):
        newP[i] = alfa[i + 1] * newP[i + 1] + betta[i + 1]

    # Check for convergence
    max_diff = np.max(np.abs(newP - oldP))

    # Update oldP for the next iteration
    oldP[:] = newP
    iter += 1

# Print results
print("Iterations:", iter)
print("X\tP\tTrue")
for i in range(n):
    x = i * dx
    true_value = np.exp(-pi * pi * dt * iter) * np.sin(pi * x)
    print(x, "\t", newP[i], "\t", true_value)

# Create the plot
x = [i * dx for i in range(n)]
y = newP
true_y = [np.exp(-pi * pi * dt * iter) * np.sin(pi * i * dx) for i in range(n)]

plt.plot(x, y, label='Numerical Solution', color='blue')
plt.plot(x, true_y, label='Analytical Solution', color='red')

# Add title and labels
plt.title("Temperature Distribution using Thomas Algorithm")
plt.xlabel("X")
plt.ylabel("P")
plt.legend()
plt.show()
#2thomas 
# import numpy as np
# import matplotlib.pyplot as plt


# n = 101
# pi = 3.14159265359
# C = 1.0

# dx = 1.0 / (n - 1)
# dt = 0.01 * dx / (2.0 * C)
# eps = 0.00001
# oldP = np.zeros(n)
# newP = np.zeros(n)
# d = np.zeros(n)
# alfa = np.zeros(n)
# betta = np.zeros(n)

# for i in range(n):
#     if i <= n // 2:
#         oldP[i] = 1.0


# a = C / dx
# b = 1.0 / dt - C / dx
# c = 0.0

# # Main iteration loop
# iter = 0
# max_diff = float('inf')

# while iter < 100 and max_diff > eps:
#     # Fill the d vector
#     for i in range(n):
#         d[i] = oldP[i] / dt

#     # Forward substitution (Thomas algorithm)
#     alfa[1] = 0.0
#     betta[1] = 1.0
#     for i in range(1, n - 1):
#         alfa[i + 1] = -a / (b + c * alfa[i])
#         betta[i + 1] = (d[i] - c * betta[i]) / (b + c * alfa[i])

    
#     newP[n - 1] = 0.0 
#     for i in range(n - 2, -1, -1):
#         newP[i] = alfa[i + 1] * newP[i + 1] + betta[i + 1]

    
#     max_diff = np.max(np.abs(newP - oldP))

    
#     oldP[:] = newP
#     iter += 1

# # Print results
# print("Iterations:", iter)
# print("VARIABLES = \"X\", \"P\"")
# print("ZONE I=", n, ", F=POINT")
# for i in range(n):
#     print(i * dx, "\t", newP[i])

# # Create the plot
# x = [i * dx for i in range(n)]
# y = newP

# plt.plot(x, y, label='graph ', color='blue')
# plt.title("Temperature Distribution using Thomas Algorithm")
# plt.xlabel("X")
# plt.ylabel("P")
# plt.legend()
# plt.show()
# #1exact
# import numpy as np
# import matplotlib.pyplot as plt

# # Define constants
# n = 51
# pi = 3.14159265359

# # Initialize parameters
# dx = 1.0 / (n - 1)
# dt = 0.5 * dx * dx
# eps = 0.00001

# # Initialize arrays
# oldP = np.zeros(n)
# newP = np.zeros(n)

# # Set initial conditions
# for i in range(n):
#     oldP[i] = np.sin(pi * i * dx)

# # Main loop
# iter = 0
# max_diff = float('inf')  # Initialize max_diff to a large value

# while max_diff > eps:
#     # Boundary conditions
#     oldP[0] = 0.0
#     newP[0] = 0.0
#     oldP[n - 1] = 0.0
#     newP[n - 1] = 0.0

#     # Update interior points
#     for i in range(1, n - 1):
#         newP[i] = oldP[i] + dt * ((oldP[i + 1] - 2.0 * oldP[i] + oldP[i - 1]) / (dx * dx))

#     # Calculate the maximum difference
#     max_diff = np.max(np.abs(newP - oldP))

#     # Update oldP
#     oldP[:] = newP

#     iter += 1

# # Print results
# print("VARIABLES = \"X\", \"P\", \"True\"")
# print("ZONE I=", n, ", F=POINT")
# for i in range(n):
#     true_value = np.exp(-pi * pi * dt * iter) * np.sin(pi * i * dx)
#     print(i * dx, "\t", newP[i], "\t", true_value)

# # Create the plot
# x = [i * dx for i in range(n)]
# y = newP
# true_y = [np.exp(-pi * pi * dt * iter) * np.sin(pi * i * dx) for i in range(n)]

# plt.plot(x, y, label='Numerical Solution', color='blue')
# plt.plot(x, true_y, label='Analytical Solution', color='red')

# # Add title and labels
# plt.title("Temperature Distribution")
# plt.xlabel("X")
# plt.ylabel("P")

# plt.legend()
# plt.show()

#2exact
# import numpy as np
# import matplotlib.pyplot as plt

# # Define constants
# n = 101
# pi = 3.1415
# C = 1.0

# # Initialize arrays
# dx = 1.0 / (n - 1)
# dt = 0.01 * dx / (2.0 * C)
# eps = 0.00001
# oldP = np.zeros(n)
# newP = np.zeros(n)

# # Set initial conditions
# for i in range(n):
#     if i <= n // 2:
#         oldP[i] = 1.0

# # Set boundary conditions
# oldP[0] = 1.0
# oldP[n - 1] = 0.0

# # Main loop
# iter = 0
# while iter < 10000:
#     newP[0] = 1.0
#     newP[n - 1] = 0.0

#     # Update interior points
#     for i in range(1, n - 1):
#         newP[i] = oldP[i] - C * dt * ((oldP[i] - oldP[i - 1]) / dx)

#     # Update max difference
#     max_diff = np.max(np.abs(newP - oldP))

#     # Update oldP
#     oldP[:] = newP

#     iter += 1

# # Print results
# print("Iterations:", iter)

# # Print output in a format similar to the C++ code
# print("VARIABLES = \"X\", \"P\"")
# print("ZONE I=", n, ", F=POINT")
# for i in range(n):
#     print(i * dx, "\t", newP[i])

# # Create the plot
# x = [i * dx for i in range(n)]
# y = newP

# plt.plot(x, y, label='Numerical Solution', color='blue')

# # Add title and labels
# plt.title("Temperature Distribution")
# plt.xlabel("X")
# plt.ylabel("P")

# # Add legend
# plt.legend()

# # Show the plot
# plt.show()

File: 5th sem/hw5.cpp
#include <iostream>
#include <cmath>
#define n 51
int main() {
    int i, iter = 0;
    double dx, dt, a, b, c, d[n], f[n], alfa[n], betta[n], newP[n], oldP[n], eps, max;
    const double pi = 3.14159265359;
    dx = 1.0 / (n - 1);
    dt = 0.5 * dx * dx;
    eps = 0.0000000001;
    for (i = 0; i < n; i++)
        oldP[i] = sin(pi * i * dx);
    for (i = 0; i < n; i++)
        f[i] = 0.0;
    a = -1.0 / (dx * dx);
    b = 1.0 / dt + 2.0 / (dx * dx);
    c = -1.0 / (dx * dx);
    do {
        for (i = 0; i < n; i++)
            d[i] = oldP[i] / dt + f[i];
        // Прямой ход метода прогонки
        alfa[1] = 0.0;
        betta[1] = 0.0;
        for (i = 1; i < n - 1; i++) {
            alfa[i+1] = -a / (b + c * alfa[i]);
            betta[i+1] = (d[i] - c * betta[i]) / (b + c * alfa[i]);
        }
        // Обратный ход метода прогонки
        newP[n - 1] = 0.0;
        for (i = n - 2; i >= 0; i--) {
            newP[i] = alfa[i+1] * newP[i + 1] + betta[i+1];
        }
        // Проверка на сходимость
        max = 0.0;
        for (i = 0; i < n; i++) {
            if (max < fabs(newP[i] - oldP[i]))
                max = fabs(newP[i] - oldP[i]);
        }
        // Обновление значений для следующей итерации
        for (i = 0; i < n; i++)
            oldP[i] = newP[i];
        iter++;
    } while (max > eps);
    std::cout << "Iterations: " << iter << std::endl;
    std::cout << "X\tP\tTrue\n";
    for (i = 0; i < n; i++) {
        double x = i * dx;
        double true_value = exp(-pi * pi * dt * iter) * sin(pi * x);
        std::cout << x << "\t" << newP[i] << "\t" << true_value << std::endl;
    }
    return 0;}
// /2. Thomas
// #include <iostream>
// #include <cmath>
// #define n 101
// int main() {
//     int i, iter = 0;
//     double dx, dt, a, b, c, d[n], alfa[n], betta[n], oldP[n], newP[n], eps, max;
//     const double pi = 3.14159265359, C = 1.0;

//     dx = 1.0 / (n - 1);
//     dt = 0.01 * dx / (2.0 * C);
//     eps = 0,00001;
//     for (i = 0; i < n; i++) {
//         oldP[i] = 0.0;
//         if (i <= n / 2) {
//             oldP[i] = 1.0;
//         }
//     }

//     a = C / dx;
//     b = 1.0 / dt - C / dx;
//     c = 0.0;

//     do {
//         for (i = 0; i < n; i++) {
//             d[i] = oldP[i] / dt;
//         }
//         // Прямой ход метода прогонки
//         alfa[1] = 0.0;
//         betta[1] = 1.0;
//         for (i = 1; i < n - 1; i++) {
//             alfa[i + 1] = -a / (b + c * alfa[i]);
//             betta[i + 1] = (d[i] - c * betta[i]) / (b + c * alfa[i]);
//         }
//         // Обратный ход метода прогонки
//         newP[n - 1] = 0.0;
//         for (i = n - 2; i >= 0; i--) {
//             newP[i] = alfa[i + 1] * newP[i + 1] + betta[i + 1];
//         }
//         // Проверка на сходимость
//         max = 0.0;
//         for (i = 0; i < n; i++) {
//             if (max < fabs(newP[i] - oldP[i])) {
//                 max = fabs(newP[i] - oldP[i]);
//             }
//         }
//         // Обновление значений для следующей итерации
//         for (i = 0; i < n; i++) {
//             oldP[i] = newP[i];
//         }
//         iter++;
//     } while (iter < 100);
//     std::cout << "Iterations: " << iter << std::endl;
//     std::cout << "VARIABLES = \"X\", \"P\"\n";
//     std::cout << "ZONE I=" << n << ", F=POINT\n";
//     for (i = 0; i < n; i++) {
//         std::cout << i * dx << " " << newP[i] << std::endl;
//     }
//     return 0;
// }


// 1 exact
// #include <iostream>
// #include <cmath>
// #define n 51
// int main() {
//     int i, iter = 0;
//     double dx, dt, oldP[n], newP[n], eps, max;
//     const double pi = 3.14159265359;
//     dx = 1.0 / (n - 1);
//     dt = 0.5 * dx * dx;
//     eps = 0.00001;
//     for (i = 0; i < n; i++) {
//         oldP[i] = sin(pi * i * dx);
//     }
//     do {
//         oldP[0] = 0.0;
//         newP[0] = 0.0;
//         oldP[n - 1] = 0.0;
//         newP[n - 1] = 0.0;
//         for (i = 1; i < n - 1; i++) {
//             newP[i] = oldP[i] + dt * ((oldP[i + 1] - 2.0 * oldP[i] + oldP[i - 1]) / (dx * dx));
//         }
//         max = 0.0;
//         for (i = 0; i < n; i++) {
//             if (max < fabs(newP[i] - oldP[i])) {
//                 max = fabs(newP[i] - oldP[i]);
//             }
//         }
//         for (i = 0; i < n; i++) {
//             oldP[i] = newP[i];
//         }
//         iter++;
//     } while (max > eps);
//     std::cout << "VARIABLES = \"X\", \"P\", \"True\"\n";
//     std::cout << "ZONE I=" << n << ", F=POINT\n";
//     for (i = 0; i < n; i++) {
//         double true_value = exp(-pi * pi * dt * iter) * sin(pi * i * dx);
//         std::cout << i * dx << "\t" << newP[i] << "\t" << true_value << std::endl;
//     }
//     return 0;
// }


//2exact
// #include <iostream>
// #include <cmath>
// #define n 101
// int main() {
//     int i, iter = 0;
//     double dx, dt, newP[n], oldP[n], eps, max;
//     const double pi = 3.1415, C = 1.0;
//     dx = 1.0 / (n - 1);
//     dt = 0.01 * dx / (2.0 * C);
//     eps = 0.00001;
//     for (i = 0; i < n; i++) {
//         oldP[i] = 0.0;
//         if (i <= n / 2) {
//             oldP[i] = 1.0;
//         }
//     }
//     do {
//         oldP[0] = 1.0; 
//         newP[0] = 1.0;
//         oldP[n - 1] = 0.0;
//         newP[n - 1] = 0.0;

//         for (i = 1; i < n - 1; i++) {
//             newP[i] = oldP[i] - C * dt * ((oldP[i] - oldP[i - 1]) / dx);
//         }
//         oldP[0] = 1.0; 
//         newP[0] = 1.0;
//         oldP[n - 1] = 0.0;
//         newP[n - 1] = 0.0;
//         max = 0.0;
//         for (i = 0; i < n; i++) {
//             if (max < fabs(newP[i] - oldP[i])) {
//                 max = fabs(newP[i] - oldP[i]);
//             }
//         }
//         for (i = 0; i < n; i++) {
//             oldP[i] = newP[i];
//         }
//         iter++;
//     } while (iter < 10000);
//     std::cout << "Iterations: " << iter << std::endl;
//     // Вывод результатов в консоль
//     std::cout << "VARIABLES = \"X\", \"P\"\n";
//     std::cout << "ZONE I=" << n << ", F=POINT\n";
//     for (i = 0; i < n; i++) {
//         std::cout << i * dx << "\t" << newP[i] << std::endl;
//     }
//     return 0;
// }
